README.md
# skewtDist

**skewtDist** is an R package for two families of skew-t distributions that have different tail behavior for left and right tails, namely the family of the Asymmetric Student t-distributions (AST) distributions introduced by Zhu and Galbraith (2010) (ZG), and the the family of generalized asymmetric t-distributions (GAT) introduced by Baker (2018) (BAK). The importance of these two families is that they go beyond the symmetric tail behaviors of the skew-t distributions derived from skew-normal distributions, as described in Azzalini and Capitanio (2014), and hence can provide better fits for certain data arising in applications, especially for asset returns.

The **skewtDist** package aims to provide a set of tools for modelling non-normally distributed data with skew-t distributions. The package includes, for both AST and GAT families, the following functionality: (a) computation of probability density functions, cumulative distribution functions, quantiles, moments (mean, variance, skewness and excess kurtosis) (b) generation of random variables (c) information matrix computation, and (d) maximum likelihood estimation skew-t distribution parameters. 

## Installation

The package may be installed and loaded with the following code.
<!---You can install the released version of st from [CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("st")
```--->

``` r
devtools::install__github("dan9401/skewtDist")
```



## Example

This is a basic example which shows you how to solve a common problem:

``` r
pars <- c(0.12, 0.6, 0.6, 3, 5)
data <- rast(1000, pars = pars)
hist(data, breaks = 50, probability = TRUE)
x <- seq(-3, 3, 0.01)
y <- dast(x, pars = pars)
lines(x, y, col = 4)
```

An example of fitting the AST MLE:

``` r
data(retSW)
plxs <- retSW['2006-01/', 1]
fit <- astMLE(plxs)
summary(fit)
plot(fit, 2)
```

For more examples, please refer to the vignette.
[Vignette](vignettes/VignetteSkewtDist.pdf)


Changelog
1. kept data/ , but have to confirm with Doug whether it can be published
2. removed etc/ , mostly irrelevant
3. removed plots/ , mostly irrelevant as well
4. in R/ , deleted all gat-*.R files; in man/ , deleted all gat-*.Rd; in test/testthat/ , deleted all test-gat-*.R
5. vignette/ is untouched for now, as Doug may want to add some finishing touch
6. edited DESCRIPTION, removed gat, needs double checking
7. leaving NAMESPACE untouched, let roxygen2 does the job
8. removing all.Rout and .RDataTmp, hardly matters
9. keeping README.md for now
10. repurpose sink-example.txt as change log

tests/testthat.R
library(testthat)
library(skewtDist)
test_check("skewtDist")


R/ast-mle.R
#' @title Fitting function for Asymmetric Student-t distribution
#'
#' @name astMLE
#' @aliases astmle
#'
#' @description Method for fitting an AST distribution to a univariate data series by Maximum Likelihood Estimation,
#' returns an \code{ast} object.
#'
#' @param data a univariate data object to be fitted
#' @param start_pars a named numeric vector of starting parameters for the optimization algorithm, not all parameters are needed
#' @param fixed_pars a named numeric vector of parameters to be kept fixed during the optimization routine, not all parameters are needed
#' @param solver solver used for MLE, one of 'nlminb', 'nloptr', 'Rsolnp', default is 'nlminb'
#' @param solver_control list of control arguments passed to the solver
#' @param symmetric a logical argument, when TRUE, the function fits an SST distribution(Symmetric Student-t, nu1 = nu2) instead, default to FALSE
#'
#' @return
#' A \code{ast} object(S3), the components of the object are:
#'     \item{data}{the univariate data object for the AST distribution to be fitted}
#'     \item{solver}{the solver called}
#'     \item{solver_control}{the list of control argumetns passed to the solver called}
#'     \item{start_pars}{named numeric vector of starting parameters used}
#'     \item{fixed_pars}{named numeric vector of fixed parameters used}
#'     \item{symmetric}{logical argument controlling the symmetry of tail parameters in the MLE}
#'     \item{solver_result}{output of the called solver}
#'     \item{fitted_pars}{named vector of fitted arguemnts of the AST distribution}
#'     \item{objective}{the optimal log-likelihood value obtained by the solver}
#'     \item{time_elapsed}{the time elapesed for the MLE routine}
#'     \item{message}{the message of convergence status produced by the called solver}
#'     \item{standard_errors}{standard errors of the fitted parameters}
#'
#' @details
#' The \code{astMLE} function fits an AST distribution to a univariate data series by estimating the distribution parameters
#' through Maximum Likelihood Estimation.
#'
#' For details of the list of control arguments, please refer to \code{nlminb}, \code{nloptr::nloptr}, \code{Rsolnp::solnp}
#'
#' @references
#' Zhu, D., & Galbraith, J. W. (2010). A generalized asymmetric Student-t distribution with application to financial econometrics. Journal of Econometrics, 157(2), 297-305.\url{https://www.sciencedirect.com/science/article/pii/S0304407610000266}
#' \url{https://econpapers.repec.org/paper/circirwor/2009s-13.htm}
#'
#' @examples
#' pars <- c(0.12, 0.6, 0.6, 6, 5)
#' data <- rast(1000, pars = pars)
#' fit <- astMLE(data)
#' summary(fit)
#' moments(fit)
#' fitted(fit)
#' fit$standard_errors
#' fit$objective
#' plot(fit, type = "density")

#' @rdname astMLE
#' @export
# fit function for ast distribution
astMLE <- function(data, start_pars = c(), fixed_pars = c(), solver = c("nlminb", "nloptr", "Rsolnp"), solver_control = list(), symmetric = FALSE) {
  if (!is.numeric(data) || length(data) == 0)
    stop("Data must be a numeric vector of non-zero length.")

  # function that checks parameter bounds, needs to be edited
  #check_bound(start_pars)
  #check_bound(fixed_pars)
  solver = match.arg(solver)
  fit <- astFit_local(data, start_pars, fixed_pars, solver, solver_control, symmetric)
  if (symmetric == TRUE) {
    standard_errors <- sqrt(diag(solve(sstInfoMat(fit$fitted_pars)))/length(data))
  } else {
    standard_errors <- sqrt(diag(solve(astInfoMat(fit$fitted_pars)))/length(data))
  }

  fit$standard_errors <- standard_errors

  structure(fit, class = "ast")
}

astFit_local <- function(data, start_pars, fixed_pars, solver, solver_control, symmetric) {
  start_time <- Sys.time()
  ipars <- i_pars(start_pars, fixed_pars, symmetric)

  fixed_pars <- ipars$fixed_pars
  if (is.null(start_pars)) {
    start_pars <- rep(NA, 5)
  } else {
    start_pars <- start_pars[ipars$name]
  }
  names(fixed_pars) <- names(start_pars) <- ipars$name
  est_idx <- which(is.na(fixed_pars))
  x0 <- ipars$start_pars[est_idx]
  lb <- ipars$lower_bound[est_idx]
  ub <- ipars$upper_bound[est_idx]
  # arglist is an argument for llast and llast_grad
  arglist <- list(data = data,
                  fixed_pars = fixed_pars)

  # will be expanded by number of solvers developed should also implement a fixed parameter version for this
  if (solver == "nloptr") {
    res <- nloptr::nloptr(x0 = x0,
                          eval_f = llast,
                          eval_grad_f = llast_grad,
                          lb = lb,
                          ub = ub,
                          opts = solver_control,
                          arglist = arglist)
    # this is temporary, fitted should be a list with its own elements
    sol_res <- res  #list(res$pars, ...)
    fitted_pars <- res$solution
    names(fitted_pars) <- ipars$name[est_idx]
    objective <- sol_res$objective
    message <- sol_res$message
  } else if (solver == "Rsolnp") {
    res <- Rsolnp::solnp(pars = x0,
                         fun = llast,
                         LB= lb,
                         UB = ub,
                         control = solver_control,
                         arglist = arglist)
    sol_res <- res  #list(res$pars, ...)
    fitted_pars <- res$pars
    names(fitted_pars) <- ipars$name[est_idx]
    objective <- sol_res$values[length(sol_res$values)]
    message <- sol_res$convergence
  } else if (solver == "nlminb") {
    res <- nlminb(start = x0,
                  objective = llast,
                  gradient = llast_grad,
                  arglist = arglist,
                  control = solver_control,
                  lower = lb,
                  upper = ub)
    sol_res <- res  #list(res$pars, ...)
    fitted_pars <- res$par
    names(fitted_pars) <- ipars$name[est_idx]
    objective <- sol_res$objective
    message <- sol_res$message
  } else {
    NA
  }
  time_elapsed <- Sys.time() - start_time

  fitted_pars <- c(fitted_pars, fixed_pars[!is.na(fixed_pars)])
  if (symmetric == TRUE) {
    fitted_pars <- fitted_pars[c("mu", "sigma", "alpha", "nu")]
  } else {
    fitted_pars <- fitted_pars[c("mu", "sigma", "alpha", "nu1", "nu2")]
  }

  list(data = data, solver = solver, solver_control = solver_control,
       start_pars = start_pars, fixed_pars = fixed_pars, symmetric = symmetric,
       solver_result = sol_res, fitted_pars = fitted_pars,
       objective = objective, time_elapsed = time_elapsed, message = message)
}

# log-likelihood function of the AST distributions pars: parameter values y: data which you fit the distribution on
llast <- function(pars, arglist) {
  y <- arglist$data
  fixed_pars <- arglist$fixed_pars

  est_idx <- which(is.na(fixed_pars))
  fix_idx <- which(!(is.na(fixed_pars)))
  all_pars <- c()
  all_pars[est_idx] <- pars
  all_pars[fix_idx] <- fixed_pars[fix_idx]

  mu <- all_pars[1]
  sigma <- all_pars[2]
  alpha <- all_pars[3]
  nu1 <- all_pars[4]
  nu2 <- ifelse(length(all_pars) == 4, nu1, all_pars[5])
  T_ <- length(y)
  y1 <- y[y <= mu]
  y2 <- y[y > mu]

  logl <- -T_ * log(sigma) - 0.5 * (nu1 + 1) * sum(log(1 + ((y1 - mu)/(2 * alpha * sigma * K(nu1)))^2/nu1)) -
    0.5 * (nu2 + 1) * sum(log(1 + ((y2 - mu)/(2 * (1 - alpha) * sigma * K(nu2)))^2/nu2))
  -logl
}

llast_grad <- function(pars, arglist) {
  y <- arglist$data
  fixed_pars <- arglist$fixed_pars

  est_idx <- which(is.na(fixed_pars))
  fix_idx <- which(!(is.na(fixed_pars)))
  all_pars <- c()
  all_pars[est_idx] <- pars
  all_pars[fix_idx] <- fixed_pars[fix_idx]

  mu <- all_pars[1]
  sigma <- all_pars[2]
  alpha <- all_pars[3]
  nu1 <- all_pars[4]
  nu2 <- ifelse(length(all_pars) == 4, nu1, all_pars[5])
  T_ <- length(y)
  y1 <- y[y <= mu]
  y2 <- y[y > mu]

  g_mu <- sum((nu1 + 1) / L(all_pars, y1) / nu1 * (y1 - mu) / (2 * alpha * sigma * K(nu1))^2) +
    sum((nu2 + 1) / R(all_pars, y2) / nu2 * (y2 - mu) / (2 * (1 - alpha) * sigma * K(nu2))^2)
  g_sigma<- -T_ / sigma + (nu1 + 1) / sigma * sum(1 - 1 / L(all_pars, y1)) + (nu2 + 1) / sigma * sum(1 - 1 / R(all_pars, y2))
  g_alpha <- (nu1 + 1) / alpha * sum(1 - 1 / L(all_pars, y1)) - (nu2 + 1) / (1 - alpha) * sum(1 - 1 / R(all_pars, y2))
  g_nu1 <- sum(- log(L(all_pars, y1)) / 2 + (nu1 + 1) / 2 * D(nu1) * (L(all_pars, y1) - 1) / L(all_pars, y1))
  g_nu2 <- sum(- log(R(all_pars, y2)) / 2 + (nu2 + 1) / 2 * D(nu2) * (R(all_pars, y2) - 1) / R(all_pars, y2))
  gradient <- -c(mu = g_mu, sigma = g_sigma, alpha = g_alpha, nu1 = g_nu1, nu2 = g_nu2)

  # gradient <- pmax(gradient, -1e6)
  # gradient <- pmin(gradient, 1e6)
  return(gradient[est_idx])
}

i_pars <- function(start_pars, fixed_pars, symmetric) {

  # insert the initial guess problem
  # possible code for using mode as the initial guess for mu
  # hist <- hist(data, breaks = 1001), only work best when a large sample of data
  # mode <- hist$mids[which(hist$counts == max(hist$counts))]
  if (symmetric == TRUE) {
    b_df <- data.frame(name = c("mu", "sigma", "alpha", "nu"),
                       lower_bound = c(-1000, 0, 0, 0),
                       upper_bound = c(1000, 1000, 1, 1000),
                       order = 1:4)
    start_pars_default <- c(mu = 0, sigma = 1, alpha = 0.5, nu = 2)
  } else {
    b_df <- data.frame(name = c("mu", "sigma", "alpha", "nu1", "nu2"),
                       lower_bound = c(-Inf, 0, 0, 0, 0),
                       upper_bound = c(Inf, Inf, 1, Inf, Inf),
                       order = 1:5)
    start_pars_default <- c(mu = 0, sigma = 1, alpha = 0.5, nu1 = 2, nu2 = 2)
  }
  start_pars <- c(start_pars, start_pars_default[!(b_df$name %in% names(start_pars))])
  sp_df <- data.frame(start_pars = start_pars,
                      name = names(start_pars))
  fp_df <- data.frame(fixed_pars = fixed_pars,
                      name = names(fixed_pars))
  if (length(fp_df != 0)) {
    p_df <- merge(sp_df, fp_df, by = "name", all = T)
  } else {
    sp_df$fixed_pars = NA
    p_df = sp_df
  }
  ipars <- merge(p_df, b_df, by = "name", all = T)
  ipars <- ipars[order(ipars$order), ]
  # return
  ipars
}


R/ast-methods.R
#' @title Methods for ast class
#'
#' @description Methods for ast S3 class
#'
#' @param object A AST fit object of class \code{\link{ast}}
#' @param x A AST fit object of class \code{\link{ast}}
#' @param method one of "numerical" and "analytical", calculating the moments using numerical integration / analytical formula
#' @param type one of "density" and "QQplot"
#' @param y not used
#' @param dist the distribution (either normal or ast)
#' @param envelope confidence interval
#' @param ... additional arguments passed to the underlying method
#'
#' @details should also add the empirical moments
#'
#' @name ast-methods
#' @aliases summary.ast
#' @aliases moments.ast
#' @aliases plots.ast
#'
#' @examples
#' pars <- c(0.12, 0.6, 0.6, 6, 5)
#' data <- rast(1000, pars = pars)
#' solver_control <- list(eval.max = 10^3, iter.max = 10^3)
#' fit <- astMLE(data, solver = 'nlminb', solver_control = solver_control)
#' summary(fit)
#' moments(fit)
#' plot(fit, type = "density")

#' @rdname ast-methods
#' @export
summary.ast <- function(object, ...) {
  dist <- ifelse(object$symmetric == TRUE, "SST", "AST")
  pars <- rbind(object$start_pars, object$fixed_pars)
  res <- rbind(object$fitted_pars, object$standard_errors)
  colnames(pars) <- colnames(res) <- names(object$fitted_pars)
  rownames(pars) <- c("start_pars", "fixed_pars")
  rownames(res) <- c("fitted_pars", "standard_errors")

  cat("Distribution: ", dist, "\n")
  cat("Observations: ", length(object$data), "\n")
  cat("\nResult:\n")
  print(res)
  cat("\nLog-likelihood", object$objective)
  cat("\n\nSolver: ", object$solver)
  cat("\n\n")
  print(pars)
  cat("\nTime elapsed: ", object$time_elapsed)
  cat("\nConvergence Message: ", object$message)
}

#' @rdname ast-methods
#' @export
moments.ast <- function(x, method = c("analytical", "numerical"), ...) {
  pars <- x$fitted_pars
  astMoments(pars = pars, method)
}

#' @rdname ast-methods
#' @export
print.ast <- function(x, ...) {
  dist <- ifelse(x$symmetric == TRUE, "SST", "AST")
  res <- rbind(x$fitted_pars, x$standard_errors)
  colnames(res) <- names(x$fitted_pars)
  rownames(res) <- c("fitted_pars", "standard_errors")

  cat("Distrifitbution: ", dist, "\n")
  cat("Observations: ", length(x$data), "\n")
  cat("\nResult:\n")
  print(res)
}

#' @rdname ast-methods
#' @export
plot.ast <- function(x, y = NULL, type = NULL, dist = "ast", envelope = 0.95, ...) {
  if (is.null(type)) {
    selection <- 1
    while (selection) {
      selection <- menu(c("Density", "qqplot"), title = "Make a plot selection (or 0 to exit)")
      if (selection == 1) {
        density_ast(x, ...)
      } else if(selection == 2) {
        qqplot_ast(x, dist, envelope = 0.95, ...)
      }
    }
  } else {
    if (type == "density") {
      density_ast(x, ...)
    } else if(type == "qqplot") {
      qqplot_ast(x, dist = dist, envelope = envelope, ...)
    }
  }
}


R/ast-dist.R
#' @title Asymmetric Student-t Distribution
#'
#' @name AST
#' @aliases ast
#' @aliases dast
#' @aliases past
#' @aliases qast
#' @aliases rast
#'
#' @description Probablity density function(PDF), Cumulative distribution function(CDF), Quantile function and Random generation of the AST distribution
#'
#' @param x,q vector of quantiles
#' @param p vector of probablilities
#' @param n number of observations for random generation
#' @param mu location parameter
#' @param s scale parameter, \eqn{s > 0}
#' @param alpha skewness parameter, \eqn{0 < alpha < 1}
#' @param nu1 degrees of freedom / tail parameter for the left tail, \eqn{ nu1 > 0}
#' @param nu2 degrees of freedom / tail parameter for the right tail, \eqn{ nu2 > 0}
#' @param pars a vector that contains mu, s, alpha, nu1, nu2, if pars is specified, mu, s, alpha, nu1, nu2 should not be specified
#'
#' @return
#' \code{dast} gives the density, \code{past} gives the distribution function, \code{qast} gives the quantile function, and \code{rast} generates random samples for AST distribution.
#'
#' @details
#' The 'asymmetric' in AST distribution, not only suggests skewness in the distribution,
#' but also the asymmetry in the two tail powers of the distribution.
#' \itemize{
#'     \item Location parameter \code{mu} is the mode, but not necessarily the mean of the distribution.
#'     \item Scale parameter \code{s} is not necessarily the standard deviation.
#'     \item The distribution skews to the right when the skewness parameter \eqn{alpha < 0.5},
#'     skews to the left when \eqn{alpha > 0.5}.
#'     \item The location paramter \code{mu} always locates at the \eqn{\alpha}-th percentile of the distribution.
#'     The two degrees of freedom / tail parameters each controls one tail of the distribution,
#'     separated at the location paramter \code{mu}.
#'     The left tail parameter \code{nu1} only affects the left half(0th to \eqn{\alpha}-th percentile) of the distribution,
#'     while the right tail paramter \code{nu2} only affects the right half(\eqn{\alpha}-th to 100-th percentile) of the distribution.
#' }
#' @importFrom graphics abline hist lines points
#' @importFrom stats cov2cor dnorm nlminb ppoints pt qnorm qt quantile rt runif
#' @importFrom utils menu
#' @importFrom DistributionUtils safeIntegrate
#' @references
#' Zhu, D., & Galbraith, J. W. (2010). A generalized asymmetric Student-t distribution with application to financial econometrics. Journal of Econometrics, 157(2), 297-305.\url{https://www.sciencedirect.com/science/article/pii/S0304407610000266}
#' \url{https://econpapers.repec.org/paper/circirwor/2009s-13.htm}
#'
#' @examples
#' # The parameter values are specially set for a volatile portfolio.
#' # density at the mu is always 1 / s
#' d <- dast(0.12, 0.12, 0.6, 0.6, 3, 5)
#' # cumulative distribution at mu is alpha
#' p <- past(0.12, 0.12, 0.6, 0.6, 3, 5)
#' # quantile at alpha is mu
#' q <- qast(0.4, 0.12, 0.6, 0.6, 3, 5)
#' data <- rast(1000, 0.12, 0.6, 0.6, 3, 5)
#' hist(data, breaks = 50, probability = TRUE)
#'
#' # using the 'pars' argument
#' pars <- c(0.12, 0.6, 0.6, 3, 5)
#' x <- seq(-3, 3, 0.01)
#' y <- dast(x, pars = pars)
#' lines(x, y, col = 4)
#'
#' pars1 <- c(0, 2, 0.3, 5, 5)
#' pars2 <- c(0, 2, 0.5, 5, 5)
#' pars3 <- c(0, 2, 0.7, 5, 5)
#' y1 <- dast(x, pars = pars1)
#' y2 <- dast(x, pars = pars2)
#' y3 <- dast(x, pars = pars3)
#' plot(x, y1, type = "l", main = expression(alpha), xlab = "x", ylab = "density")
#' lines(x, y2, lty = 2)
#' lines(x, y3, lty = 3)
#' abline(v = 0, col = 4, lty = 2)
#' legend(x = "topleft", legend = c("alpha = 0.3", "alpha = 0.5", "alpha = 0.7"), lty = 1:3)

#' @rdname AST
#' @export
dast <- function(x, mu = 0, s = 1, alpha = 0.5, nu1 = Inf, nu2 = Inf, pars = NULL) {
    if (!is.null(pars)) {
      if (!missing(mu)) {
        stop("Only one of [mu, s, alpha, nu1, nu2] and pars needs to be specified")
      }
      mu <- pars[1]
      s <- pars[2]
      alpha <- pars[3]
      nu1 <- pars[4]
      nu2 <- pars[5]
    }
    if (!is.numeric(x))
      stop("x must be numeric")
    x1 <- x[x <= mu]
    x2 <- x[x > mu]
    d <- numeric(length(x))
    # refer to helper_functions.R for K(.)
    d[x <= mu] <- (1 + ((x1 - mu)/(2 * alpha * s * K(nu1)))^2/nu1)^(-0.5 * (nu1 + 1))/s
    d[x > mu] <- (1 + ((x2 - mu)/(2 * (1 - alpha) * s * K(nu2)))^2/nu2)^(-0.5 * (nu2 + 1))/s
    d
}

#' @rdname AST
#' @export
past <- function(q, mu = 0, s = 1, alpha = 0.5, nu1 = Inf, nu2 = Inf, pars = NULL) {
    if (!is.numeric(q))
        stop("q must be numeric")
    if (!is.null(pars)) {
      if (!missing(mu)) {
        stop("Only one of [mu, s, alpha, nu1, nu2] and pars needs to be specified")
      }
      mu <- pars[1]
      s <- pars[2]
      alpha <- pars[3]
      nu1 <- pars[4]
      nu2 <- pars[5]
    }
    B <- alpha * K(nu1) + (1 - alpha) * K(nu2)
    q <- (q - mu) / (s * B)
    # refer to helper_functions.R for K(.)
    alpha_star <- alpha * K(nu1)/ B
    # return value
    2 * alpha * pt(pmin(q, 0)/(2 * alpha_star), nu1) + 2 * (1 - alpha) * (pt(pmax(q, 0)/(2 * (1 - alpha_star)), nu2) - 0.5)
}

#' @rdname AST
#' @export
qast <- function(p, mu = 0, s = 1, alpha = 0.5, nu1 = Inf, nu2 = Inf, pars = NULL) {
    if (!is.numeric(p))
        stop("p must be numeric")
    if (!is.null(pars)) {
      if (!missing(mu)) {
        stop("Only one of [mu, s, alpha, nu1, nu2] and pars needs to be specified")
      }
      mu <- pars[1]
      s <- pars[2]
      alpha <- pars[3]
      nu1 <- pars[4]
      nu2 <- pars[5]
    }
    # refer to helper_functions.R for K(.)
    B <- alpha * K(nu1) + (1 - alpha) * K(nu2)
    alpha_star <- alpha * K(nu1)/ B
    # return value
    mu + s * B * (2 * alpha_star * ( qt(pmin(p, alpha)/(2 * alpha), nu1)) + 2 * (1 - alpha_star) * ( qt((pmax(p, alpha) + 1 - 2 * alpha)/(2 * (1 - alpha)), nu2)))
}

#' @rdname AST
#' @export
rast <- function(n, mu = 0, s = 1, alpha = 0.5, nu1 = Inf, nu2 = Inf, pars = NULL) {
    if (n < 0)
        stop("n must be non-negative")
    # refer to helper_functions.R for K(.)
    if (!is.null(pars)) {
      if (!missing(mu)) {
        stop("Only one of [mu, s, alpha, nu1, nu2] and pars needs to be specified")
      }
      mu <- pars[1]
      s <- pars[2]
      alpha <- pars[3]
      nu1 <- pars[4]
      nu2 <- pars[5]
    }
    alpha_star <- alpha * K(nu1)/(alpha * K(nu1) + (1 - alpha) * K(nu2))
    u <- runif(n)
    t1 <- rt(n, nu1)
    t2 <- rt(n, nu2)
    x <- alpha_star * abs(t1) * (sign(u - alpha) - 1) + (1 - alpha_star) * abs(t2) * (sign(u - alpha) + 1)
    # return value
    # important note here!  s has been transformed
    mu + x * s * (alpha * K(nu1) + (1 - alpha) * K(nu2))
}
